lib/playlist-generation/compute-score.ts

```diff
@@ -0,0 +1,28 @@
+import { Song } from '../song';
+import { AudioFeatures } from '../spotify';
+import { THEME, theme2Conditions, ThemeConditions } from './themes';
+
+export type Song2Score = {song: Song, score: number};
+
+export const compareSongScores = (a: Song2Score, b: Song2Score): number => a.score < b.score ? -1 : 1;
+
+type AudioFeatureEntry = [feature: keyof AudioFeatures, condition: ThemeConditions];
+
+export function computeScore(af: AudioFeatures, theme: THEME): number {
+  const conditions = theme2Conditions[theme];
+
+  const isQualifying = !Object.entries(conditions).find(([feature, condition]: AudioFeatureEntry) => {
+    if (condition.weight === 0) return false;
+    return af[feature] < condition.min || condition.max < af[feature];
+  });
+
+  if (!isQualifying) return 0;
+
+  const rawScore = Object.entries(conditions).reduce((acc, [feature, {target, weight}]: AudioFeatureEntry) => {
+    return acc + (1 - Math.abs(target - af[feature])) * weight;
+  }, 0);
+
+  const numWeighted = Object.values(conditions).reduce((acc, {weight}) => acc + +(!!weight), 0);
```

Comment:
is the `acc + +(!!weight)` a typo? Should it just be one plus sign?

Reply:
the second plus is to convert the boolean to a number. Basically I want to count all the weighted features. So the first `!` allows me to turn the weight into a boolean of either 0 or 1 depending on whether weight is not 0 or is. And then the second `!` keeps it a boolean and reverts it.